samples/wifi/provisioning/implementation_details.rst

```diff
@@ -0,0 +1,313 @@
+.. _wifi_provisioning_details:
+
+Implementation Details
+######################
+
+.. contents::
+   :local:
+   :depth: 2
+
+To better understand how the sample works, the implementation details is provided. User can modify and extend the functionality
+based on existing implementation.
+
+Task and Event Handling
+=======================
+
+To exchange command and response, we need to define the format of message.
+In this implementation, Protocol Buffers is used to achieve the goal.
+It plays two roles. First, we define the format of message in language of
+Protocol Buffers, and the outcome is .proto files. Second, Protocol Buffers
+provides an unified message encoding/decoding scheme, so it allows us to
+exchange platform-independent byte stream between device and configurator,
+on which different programming language can be used.
+
+The message Info is to allow the configurator to know the capability of device
+before actual provisioning process happens. The configurator should read
+this message before send any other message. Currently, there is only a field
+of version inside.
+
+The messages Request and Response are the main messages we use in provisioning
+process. The Request message contains command and configuration. The device is
+supposed to take proper action based on the command. If configuration is
+provided, the device should also apply. The Response message is used to reply
+to the Request. It has the command it is responding, a field that indicates
+the result of related operation, and other data if any. The Request-Response
+pair is exchanged in asynchronous matter. It means that as long as the Request
+is decoded and parsed, and corresponding task is dispatched, the Response message
+is sent immediately based on the result. The data associated with this command
+may be sent to configurator after the Response message. Currently, we defined
+5 commands.
+
```

Comment:
Insert "Operations" heading here

Reply:
My suggestion for this section is something like this. Unordered list with introductory text, summarize common responses that aren't particular to a command or operation, and cut out a lot of the text to really distill the most relevant info down to a list.

I'm much more familiar with Markdown, so I've written a mockup:

## Operations

This section explains operations and responses according to each command.

The following response statuses apply to all operations:

* `STATUS_SUCCESS`
* `STATUS_INVALID_ARGUMENT`
* `STATUS_INTERNAL_ERROR`

The following list details the operations by command and specific responses associated with each:

* `START_SCAN`: Request and receive information about access points (APs), each in a result message.
   * You can define aspects of the scan using `scan_params`. (need info on what the parameters are and how to pass them)
   * Access point information is recorded in `scan_record`. 
* `SET_CONFIG`: Receive a result message when the Wi-Fi connection status changes.
* `GET_STATUS`: Request and retrieve device status information. The device records this information in the following fields:
    * `state`: the Wi-Fi connection state.
    * `provisioning_info`: ... and so on